app/liquid_tags/youtube_tag.rb

```diff
@@ -32,42 +35,41 @@ def render(_context)
   private
 
   def parse_id_or_url(input)
-    if (input =~ REGISTRY_REGEXP)&.zero?
-      extract_youtube_id(input)
-    else
-      input_no_space = input.delete(" ")
-      raise StandardError, "Invalid YouTube ID" unless valid_id?(input_no_space)
-      return translate_start_time(input_no_space) if input_no_space.include?("?t=")
-
-      input_no_space
-    end
-  end
+    match = pattern_match_for(input)
+    raise StandardError, "Invalid YouTube ID" unless match
+
+    video_id       = match[:video_id]
+    time_parameter = match[:time_parameter]
 
-  def extract_youtube_id(url)
-    url = url.gsub(/(>|<)/, "").split(%r{(vi/|v=|/v/|youtu\.be/|/embed/)})
-    raise StandardError, "Invalid YouTube URL" if url[2].nil?
+    return video_id if time_parameter.blank?
 
-    id = url[2].split(/[^a-zA-Z0-9_-]/i) # tweak this to allow for time, fix youtube_tag_spec
-    id[0]
+    translate_start_time(video_id, time_parameter)
   end
 
-  def valid_id?(id)
-    id.match?(/\A[a-zA-Z0-9_-]{11}((\?t=)?(\d{1,}h?)?(\d{1,2}m)?(\d{1,2}s)?){5,11}?\Z/)
+  def pattern_match_for(input)
+    match = nil
+    REGEXP_OPTIONS.each do |regex|
```

Comment:
otherwise - I think you may want to use `detect`, which would do the "does it match the first one? return, try the second one, does it match? return" pattern you're implementing here... I think this is the same as what you're manually doing via each/test/break:

```ruby
REGEXP_OPTIONS.detect {|r| input.match(r) }
```

Reply:
Alas, `detect` doesn't work, as we want the results of the match not the REGEXP that matched.  I really want `detect_map` in this case.